[Touch event regions] are not generated for `touchforcechange`, all touch events are synchronous for elements with `touchcancel` listener

lilyspiniolas opened the pull request:
#### 6303013b4c9cb1e0543063d2acc386b7eb1d4a1d
<pre>
[Touch event regions] are not generated for `touchforcechange`, all touch events are synchronous for elements with `touchcancel` listener
<a href="https://bugs.webkit.org/show_bug.cgi?id=303403">https://bugs.webkit.org/show_bug.cgi?id=303403</a>
<a href="https://rdar.apple.com/165708004">rdar://165708004</a>

Reviewed by Abrar Rahman Protyasha.

In `eventTypeForEventListenerType(...)`, we previously mapped
`EventListenerRegionType::NonPassiveTouchCancel`
to `EventTrackingRegionsEventType::Touchforcechange`, causing cascading
effects which eventually lead to
`internals().touchEventTracking.touchForceChangedTracking` getting set
to synchronous. Then, in `WebPageProxy::touchEventTrackingType`, the global
tracking type is unconditionally merged with `touchForceChangedTracking`,
causing all touch events to be synchronous.

Fix this bug by removing the bad mapping. Remove
`EventListenerRegionType::NonPassiveTouchCancel` because we don&apos;t
need to generate synchronous touch cancel regions in the first place.
`touchcancel` is never cancelable since it is a notification that the
touch event has already been canceled. This matches the behavior when
touch event regions are disabled.

Add support for new EventListenerRegionType
`EventListenerRegionType::NonPassiveTouchForceChange` and
`EventListenerRegionType::TouchForceChange` so that we can properly
track touch event regions for elements with `touchforcechange`
listeners.

* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::eventTypeForEventListenerType):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::patternForEventListenerRegionType):
(WebCore::RenderLayerBacking::paintDebugOverlays):
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::computeEventListenerRegionTypes):

Canonical link: <a href="https://commits.webkit.org/303895@main">https://commits.webkit.org/303895@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f48664f46d480271d02c1e7dd46e987f8e35ae14

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133867 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6378 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45058 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141444 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85927 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/72d8d630-c966-4945-992c-b0a2b9ff9268) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135737 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6907 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6242 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102427 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/38012fe4-6a1d-40f0-ac37-dfd134646ca4) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136814 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4866 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120011 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83226 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/08afbda1-9b23-4c43-964f-f271346d467f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4745 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2365 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113899 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38127 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144090 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6048 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38706 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110777 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6130 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5154 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110978 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28155 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4600 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116265 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59795 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6100 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34540 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5946 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69564 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6191 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6054 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -384,7 +387,6 @@ OptionSet<EventListenerRegionType> Adjuster::computeEventListenerRegionTypes(con
         types.add(EventListenerRegionType::TouchEnd);
         types.add(EventListenerRegionType::TouchMove);
         types.add(EventListenerRegionType::TouchStart);
-        types.add(EventListenerRegionType::NonPassiveTouchCancel);
```
do we want to include touchforcechange here?

lilyspiniolas replied:
Yes, thank you!

lilyspiniolas committed:
[Touch event regions] are not generated for `touchforcechange`, all touch events are synchronous for elements with `touchcancel` listener
https://bugs.webkit.org/show_bug.cgi?id=303403
rdar://165708004

Reviewed by Abrar Rahman Protyasha.

In `eventTypeForEventListenerType(...)`, we previously mapped
`EventListenerRegionType::NonPassiveTouchCancel`
to `EventTrackingRegionsEventType::Touchforcechange`, causing cascading
effects which eventually lead to
`internals().touchEventTracking.touchForceChangedTracking` getting set
to synchronous. Then, in `WebPageProxy::touchEventTrackingType`, the global
tracking type is unconditionally merged with `touchForceChangedTracking`,
causing all touch events to be synchronous.

Fix this bug by removing the bad mapping. Remove
`EventListenerRegionType::NonPassiveTouchCancel` because we don't
need to generate synchronous touch cancel regions in the first place.
`touchcancel` is never cancelable since it is a notification that the
touch event has already been canceled. This matches the behavior when
touch event regions are disabled.

Add support for new EventListenerRegionType
`EventListenerRegionType::NonPassiveTouchForceChange` and
`EventListenerRegionType::TouchForceChange` so that we can properly
track touch event regions for elements with `touchforcechange`
listeners.

* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::eventTypeForEventListenerType):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::patternForEventListenerRegionType):
(WebCore::RenderLayerBacking::paintDebugOverlays):
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::computeEventListenerRegionTypes):

Canonical link: https://commits.webkit.org/303895@main
